hathor/p2p/sync_v2/transaction_streaming_client.py

```diff
@@ -87,34 +92,91 @@ def handle_transaction(self, tx: BaseTransaction) -> None:
+            return
+        self._waiting_for.remove(tx.hash)
+
+        for dep in self.get_missing_deps(tx):
+            self.log.debug('adding dependency', tx_id=tx.hash.hex(), dep=dep.hex())
+            self._waiting_for.add(dep)
+
+        self._db[tx.hash] = tx
+        self._buffer.append(tx.hash)
+
+        if not self._waiting_for:
+            self.log.debug('no pending dependencies, processing buffer')
+            self._execute_and_prepare_next()
+        else:
+            self.log.debug('pending dependencies', counter=len(self._waiting_for))
+
+        if self._tx_received % 100 == 0:
+            self.log.debug('tx streaming in progress', txs_received=self._tx_received)
+        return
+
         assert tx.hash is not None
         is_duplicated = False
         if self.partial_vertex_exists(tx.hash):
-            # We reached a block we already have. Skip it.
+            # We reached a tx we already have. Skip it.
             self._tx_repeated += 1
             is_duplicated = True
             if self._tx_repeated > self.max_repeated_transactions:
                 self.log.debug('too many repeated transactions received', total_repeated=self._tx_repeated)
                 self.fails(TooManyRepeatedVerticesError())
+        if is_duplicated:
+            pass
 
-        try:
-            # this methods takes care of checking if the block already exists,
-            # it will take care of doing at least a basic validation
-            if is_duplicated:
-                self.log.debug('tx early terminate?', tx_id=tx.hash.hex())
-            else:
-                self.log.debug('tx received', tx_id=tx.hash.hex())
-            self.sync_agent.on_new_tx(tx, propagate_to_peers=False, quiet=True, reject_locked_reward=True)
-        except HathorError:
-            self.fails(InvalidVertexError())
-            return
-        else:
-            # XXX: debugging log, maybe add timing info
-            if self._tx_received % 100 == 0:
-                self.log.debug('tx streaming in progress', txs_received=self._tx_received)
+    def get_missing_deps(self, tx: BaseTransaction) -> Iterator[bytes]:
+        for dep in tx.get_all_dependencies():
+            if self.tx_storage.transaction_exists(dep):
+                continue
+            if dep in self._db:
+                continue
+            yield dep
 
     def handle_transactions_end(self, response_code: StreamEnd) -> None:
         """This method is called by the sync agent when a TRANSACTIONS-END message is received."""
         if self._deferred.called:
             return
         self._deferred.callback(response_code)
+
+    def _execute_and_prepare_next(self) -> None:
+        self._buffer.sort(key=lambda vertex_id: self._db[vertex_id].timestamp)
+        for vertex_id in self._buffer:
+            tx = self._db[vertex_id]
+            try:
+                self.manager.on_new_tx(tx, propagate_to_peers=False, fails_silently=False)
```

Comment:
It cannot verify reward lock.

Reply:
We can test if it fully syncs. As far as we know there isn't a practical case that would break it. But _not_ skipping the reward lock here could cause sync issues if a breaking case appears.

Reply:
I agree. But I think we can do it in a separate PR because this one got too big already.